src/effects/VST3/VST3Effect.cpp

```diff
@@ -339,32 +339,129 @@ bool VST3Effect::SupportsAutomation() const
    return false;
 }
 
-bool VST3Effect::SaveSettings(
-   const EffectSettings &, CommandParameters & parms) const
+struct VST3Effect::ParameterInfo
 {
-   if(mEditController == nullptr)
+   Steinberg::Vst::ParameterInfo mVstParamInfo;
+};
+
+
+bool VST3Effect::ForEachParameter(ParameterVisitor visitor) const
+{
+   if (mEditController == nullptr)
       return false;
 
-   using namespace Steinberg;
-   
-   for(int i = 0, count = mEditController->getParameterCount(); i < count; ++i)
+   for (int i = 0, count = mEditController->getParameterCount(); i < count; ++i)
    {
-      Vst::ParameterInfo parameterInfo { };
-      if(mEditController->getParameterInfo(i, parameterInfo) == kResultOk)
+      ParameterInfo pi;
+
+      Steinberg::Vst::ParameterInfo vstParameterInfo{ };
+      if (mEditController->getParameterInfo(i, vstParameterInfo) == Steinberg::kResultOk)
       {
-         if(parameterInfo.flags & Vst::ParameterInfo::kCanAutomate)
+         if (vstParameterInfo.flags & Steinberg::Vst::ParameterInfo::kCanAutomate)
          {
-            parms.Write(
-               VST3Utils::MakeAutomationParameterKey(parameterInfo),
-               mEditController->getParamNormalized(parameterInfo.id)
-            );
+            if (!visitor(ParameterInfo{ vstParameterInfo } ))
+               return false;
          }
+         else
+         {
+            return false;
+         }
+      }
+      else
+      {
+         return false;
```

Comment:
As @vsverchinsky explained, skip over certain parameters that don't automate; but don't stop the visit or return a failure code.

Reply:
done in commit `correct the behavior of ForEachParameter...`